src/Cms/Environment.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace Kirby\Cms;
+
+use Kirby\Exception\InvalidArgumentException;
+use Kirby\Filesystem\F;
+use Kirby\Http\Server;
+use Kirby\Http\Uri;
+
+/**
+ * The environment object takes care of
+ * secure host and base URL detection, as
+ * well as loading the dedicated
+ * environment options.
+ *
+ * @package   Kirby Cms
+ * @author    Bastian Allgeier <bastian@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ */
+class Environment
+{
+    /**
+     * @var string
+     */
+    protected $root;
+
+    /**
+     * @var \Kirby\Http\Uri
+     */
+    protected $uri;
+
+    /**
+     * @param string $root
+     * @param bool|string|array|null $allowed
+     */
+    public function __construct(string $root, $allowed = null)
+    {
+        $this->root = $root;
+
+        // the current URL should be detected via possibly insecure HOST headers
+        if ($allowed === true || $allowed === null) {
+            Server::hosts(['*']);
+            $this->uri = Uri::index();
+
+        // the current URL should be detected via the server name
+        } elseif ($allowed === false) {
+            Server::hosts([]);
+            $this->uri = Uri::index();
+
+        // the current URL is predefined and not detected automatically
+        } elseif (is_string($allowed) === true) {
+            $this->uri = new Uri($allowed);
+            Server::hosts([$this->uri->host()]);
+
+        // the current URL should be auto detected from a host allowlist
+        } elseif (is_array($allowed) === true) {
+            foreach ($allowed as $url) {
+                $host = (new Uri($url))->host();
+                $hosts[]    = $host;
+            }
+
+            Server::hosts($hosts);
+            $this->uri = Uri::index();
+        } else {
+            throw new InvalidArgumentException('Invalid allow list setup for base URLs');
+        }
```

Comment:
Shouldn't we throw the exception if it doesn't match any url?

````php
if ($allowed === true || $allowed === null) {
    //
} elseif ($allowed === false) {
    //
} elseif (is_string($allowed) === true) {
    //
} elseif (is_array($allowed) === true) {
    //
}

throw new InvalidArgumentException('Invalid host!');
````

Also when no url is matched, the system works even without a host, but what exactly does allowed urls do? I didn't understand 🤔

Reply:
Then, as far as I understand it's just about prefixing URLs with correct hostnames. Not blocking access to invalid hosts. Do I understand correctly?

Reply:
You need the else statement since we are not returning anything in that method. In your example, @afbora, every setup would still end up at `throw...`.

Reply:
@distantnative I know, mine was just one such example. Will we allow the user to access the site if it doesn't match any url I'm trying to describe? Like your said in nolt:


> Q: What happens if an array of URLs is configured but none of them match the current host?
> A: Probably better to throw an exception as clearly a for the developer unintended situation appeared

Reply:
I was really wondering about this during the implementation as well. We could always fall back to simple relative URLs instead of throwing an error. Do you think this could go wrong or would be worse?

Reply:
I actually like your implementation. As I'm not sure, just wanted to point this out for security reasons. I think @lukasbestle knows these well.

Reply:
Right now you cannot get an insecure host with this PR unless you actively allow it in your config. Otherwise you will always end up with an empty string. I'm pretty sure we are fine with this, but it would indeed be good to get a second or even third opinion.

Reply:
My thinking: if someone explicitly configured something and it doesn't match with reality, an error should be thrown. That makes more sense than trying to make things "just work" for a setup that the developer explicitly didn't plan to support.

Reply:
You are right. We should keep the "just works" version when no url option is set and otherwise throw an error

Reply:
I guess then it looks good to me

Reply:
@distantnative Really sorry. Did I misunderstand? Now an exception is thrown when the type (bool, string, array) does not match (not the url) 🤔

Reply:
True we would need to throw another one here https://github.com/getkirby/kirby/pull/4173/files#diff-7c2c2d4cd5cb7754713993a0d4da17c51bbe03fd23fef782e9614c30ac1ed5e8R127 right?